crates/iota-e2e-tests/tests/grpc/v0/ledger_service/get_objects.rs

```diff
@@ -0,0 +1,436 @@
+
+// Generate the FieldPresenceChecker implementation for Object
+impl_field_presence_checker!(Object, {
+    "reference" => reference,
+    "bcs" => bcs,
+});
+
+// Generate the FieldPresenceChecker implementation for ObjectReference
+impl_field_presence_checker!(ObjectReference, {
+    "object_id" => object_id,
+    "version" => version,
+    "digest" => digest,
+});
+
+/// Helper function to assert object fields based on expected fields
+fn assert_object_fields(
+    object: &Object,
+    expected_top_level: &[&str],
+    expected_reference: &[&str],
+    scenario: &str,
+) {
+    // Check top-level fields
+    assert_field_presence(object, expected_top_level, scenario);
+
+    // If reference is expected, check nested fields
+    if expected_top_level.contains(&"reference") {
+        if let Some(ref reference) = object.reference {
+            assert_field_presence(reference, expected_reference, scenario);
+        } else {
+            panic!("Expected reference field in {scenario}, but it was None");
+        }
+    }
+}
+
+async fn assert_get_objects_request(
+    client: &mut LedgerServiceClient<tonic::transport::Channel>,
+    requests: Vec<ObjectRequest>,
+    read_mask: Option<FieldMask>,
+    max_message_size_bytes: Option<u32>,
+    expected_top_level: &[&str],
+    expected_reference: &[&str],
+    expected_has_next: bool,
+    scenario: &str,
+) -> GetObjectsResponse {
+    let request = GetObjectsRequest {
+        requests: Some(ObjectRequests { requests }),
+        read_mask,
+        max_message_size_bytes,
+    };
+
+    let mut stream = client.get_objects(request).await.unwrap().into_inner();
+
+    let response = stream.next().await.unwrap().unwrap();
+
+    // Assert all returned objects have the expected fields
+    for (idx, obj_result) in response.objects.iter().enumerate() {
+        let object = obj_result.object();
+        assert_object_fields(
+            object,
+            expected_top_level,
+            expected_reference,
+            &format!("{scenario} (object {idx})"),
+        );
+    }
+
+    // Verify has_next value
+    assert_eq!(
+        response.has_next, expected_has_next,
+        "{scenario}: has_next should be {expected_has_next}"
+    );
+
+    // If has_next is false, verify stream is exhausted
+    if !expected_has_next {
+        assert!(
+            stream.next().await.is_none(),
+            "{scenario}: stream should be exhausted when has_next is false"
+        );
+    }
+
+    response
+}
+
+#[sim_test]
+async fn get_objects_readmask_scenarios() {
+    let test_cluster = TestClusterBuilder::new()
+        .with_fullnode_enable_grpc_api(true)
+        .build()
+        .await;
+
+    let mut grpc_client = LedgerServiceClient::connect(test_cluster.grpc_url())
+        .await
+        .unwrap();
+
+    let object_id = ObjectID::from_hex_literal("0x5").unwrap().to_string();
+
+    // Table-driven tests for single-object readmask scenarios
```

Comment:
Table-driven tests sound a bit like an AI comment.
```suggestion
    // Tests for single-object readmask scenarios
```